recipe-server/normandy/recipes/api/v2/views.py

```diff
@@ -203,3 +205,91 @@ def close(self, request, pk=None):
         approval_request = self.get_object()
         approval_request.close()
         return Response(status=status.HTTP_204_NO_CONTENT)
+
+
+class IdenticonView(views.APIView):
+    @staticmethod
+    def hex_luma(hex):
+        # hex color code to RGB
+        hex = hex[1:]  # remove '#'
+        rgb = [int(h, 16) for h in [hex[i: i + 2] for i in range(0, len(hex), 2)]]
+
+        # Determine color luminance based on perceived values.
+        return (0.2126 * rgb[0]) + (0.7152 * rgb[1]) + (0.0722 * rgb[2])
+
+    def pick_pair(self, genome, base):
+        base_luminance = self.hex_luma(base)
+        good_pairs = list(filter(lambda c: abs(base_luminance - self.hex_luma(c)) > 75,
+                                 genome.colors))
+        assert len(good_pairs) > 0, 'no colors satisfy the luminance requirement'
+        return genome.choice(good_pairs, 'counter color')
+
+    def get(self, request, *, generation, seed):
+        if generation != 'v1':
+            return Response(
+                {'error': 'Invalid identicon generation, only v1 is supported.'},
+                status=status.HTTP_400_BAD_REQUEST)
+
+        genome = Genome.generate(seed)
+
+        treatments = [
+            {'name': 'SingleColor', 'weight': 1},
+            {'name': 'TwoColor', 'weight': 4},
+            {'name': 'Stripes', 'weight': 6},
+        ]
+        treatment = genome.weightedChoice(treatments, 'shield treatment')['name']
+        emoji = genome.emoji('shield emoji')
+        field_color = genome.color('field color')
+        pattern_color = self.pick_pair(genome, field_color)
+
+        context = {
+            'treatment': treatment,
+            'emoji': emoji,
+            'field_color': field_color,
+            'pattern_color': pattern_color,
+        }
+
+        # Add treatment-specific information to context
```

Comment:
Instead of a big if statement based off the treatment names, we could make the treatment values functions that take the `genome` as an argument. Then we can define them statically outside of this function and call them instead of having all of their logic here.